src/hardware/flash.rs

```diff
@@ -45,3 +63,28 @@ impl embedded_storage::nor_flash::NorFlash for Flash {
         bank.write(offset, bytes)
     }
 }
+
+impl embedded_storage_async::nor_flash::NorFlash for Flash {
+    const WRITE_SIZE: usize =
+        stm32h7xx_hal::flash::UnlockedFlashBank::WRITE_SIZE;
+    const ERASE_SIZE: usize =
+        stm32h7xx_hal::flash::UnlockedFlashBank::ERASE_SIZE;
+
+    async fn erase(&mut self, from: u32, to: u32) -> Result<(), FlashError> {
+        let mut bank = self.0.unlocked();
+        bank.erase(from, to)
+    }
+
+    async fn write(
+        &mut self,
+        offset: u32,
+        bytes: &[u8],
+    ) -> Result<(), FlashError> {
+        let mut bank = self.0.unlocked();
+        bank.write(offset, bytes)
+    }
+}
+
+// Note: While this is supported, multiple writes to the same word can trigger ECC errors due to
+// incorrect writes.
+impl embedded_storage_async::nor_flash::MultiwriteNorFlash for Flash {}
```

Comment:
Alright, so I've isolated the problem with this PR.

The comment here is indeed the problem. The datasheet specifies:
> It is not recommended to overwrite a flash word that is not virgin. The result may lead to an
inconsistent ECC code that will be systematically reported by the embedded flash memory,
as described in Section 4.7.7: Error correction code error (SNECCERR/DBECCERR)

[Datasheet](https://www.st.com/resource/en/reference_manual/rm0433-stm32h742-stm32h743753-and-stm32h750-value-line-advanced-armbased-32bit-mcus-stmicroelectronics.pdf) Section 4.3.9

My hope was that we could always set any bits from 1->0 without triggering the ECC faults. However, this appears not to be the case. I performed some manual write testing during the Stabilizer `hardware::setup` process doing the following:
1. Erase flash sector starting at 0x8100_0000 (Flash Bank 2, sector 1)
2. Write 0xAA to a full flash page (256 bits, 32 bytes)

At this point, I tried a number of test cases with varying results:
1. Write 0x00 to the entire 256 bit flash page previously written with 0xAA
    * This is acceptable and does not cause ECC faults
 2. Write 0xAA to the entire 256 bit flash page again (redundantly writing)
    * This is acceptable and does not cause ECC faults
 3. Writing 0x00 to the first byte, and writing 0xAA to the remaining bytes
    * **This triggers an ECC fault**
 4. Writing 0x00 to the first two bytes, writing 0xAA to remaining bytes
    * **Thsi triggers an ECC fault**
 5. Writing 0x00 to the first 4 bytes, writing 0xAA to remaining bytes
    * This is acceptable and does not cause ECC faults

---

Currently, `sequential-storage` "erases" an element by clearing the CRC (writing it to zero) from an item header. An item header is composed of a CRC (16 bit) + Length (16-bit) forming a 32-bit word. Thus, this operation is equivalent to test 4 above, which was observed to trigger ECC faults.

Reply:
Ah, I believe section 4.3.12 explains why this behavior is being observed:
> During each program operation, a 10- bit ECC code is associated to each 256-bit data flash
word, and the resulting 266-bit flash word information is written in non-volatile memory

I suspect the ECC code can _also_ only be written from 1->0. Based on how we manipulate the flash page, some of the ECC bits may go from 0->1 even if the original flash word only has bits going from 1->0

Reply:
Yeah, that's why it requires the trait bound `MultiwriteNorFlash`. I was hoping we could safely implement the trait, but we can't. I removed the implementation.